### ISSUE TITLE
IIS AspNetCoreModuleV2: random/endless recycles

### ISSUE THREAD

**Issue by mus65** [dotnet/aspnetcore#54531]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

After deploying an update of our application at one of our customers in IIS, our application was stuck in an endless loop. It would  always start successfully and then be recycled a few seconds later.

In the event log we saw the following:

```
An ISAPI reported an unhealthy condition to its worker process. Therefore, the worker process with process id of '9092' serving application pool '<PoolName>' has requested a recycle.
```

This only stopped after we completely restarted the IIS server. This was not reproducible at first, but happened again a few weeks later after another update.

After some investigation, we noticed that this would still happen randomly a few times a day, even after fixing the "endless loop" by restarting IIS. All auto-recycle settings on the application pool are disabled. After enabling the debug logging for the ASP .NET Core Module, we saw the following log event when a random recycle happened:

```
[2024-03-13T12:40:45.437Z, PID: 11716] [aspnetcorev2.dll] ASPNET_CORE_GLOBAL_MODULE::OnGlobalConfigurationChange ''
[2024-03-13T12:40:45.439Z, PID: 11716] [aspnetcorev2.dll] Stopping application '/LM/W3SVC/2/ROOT/<path to app>'
```

Relevant code here: 

https://github.com/dotnet/aspnetcore/blob/268a2dfc29b33e3fdb73cbac6eb198c05314d77e/src/Servers/IIS/AspNetCoreModuleV2/AspNetCore/globalmodule.cpp#L52-L68

Notice that path in the OnGlobalConfigurationChange log message is empty (as far as I can see it can't be NULL since then the recycle wouldn't be triggered). Our current assumption is that this has something to do with the problem since we don't know when/how this happens. When you edit an IIS configuration file, this contains the path to the relevant config section.

After further investigation we also saw these random recycles on other customer systems in different infrastructures. However the endless recycle loop only happened on a particular system until now. Unfortunately we have no Debug Logs for the endless loop case and can't reproduce it again. We assume that the endless loop was also caused by configuration changes with an empty path, but we can't be sure.

We currently worked around this setting [disallowRotationOnConfigChange to true](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/advanced?view=aspnetcore-8.0#disallow-rotation-on-config). We have not seen this happen since.

So the questions are:
 - what triggers these global configuration events with the empty path? The [API doc](https://learn.microsoft.com/en-us/iis/web-development-reference/native-code-api-reference/iglobalconfigurationchangeprovider-getchangepath-method) gives no indication what an empty path means.
 - Should the ASP .NET Core Module maybe ignore the configuration change if the path is empty? From my understanding this would "fix" the issue.
 
This happened with ASP .NET Core 8.0.2 on Windows Server 2016 with a self-contained x64 app.

### .NET Version

8.0.2

**Comment by ryanbuening** [comment]
Thank you for creating this ticket! I've been banging my head against the wall trying to figure out why this has been happening. We are seeing this as well in our various Blazor Server .NET 8 applications hosted on IIS. We have been seeing this more randomly, not an endless loop.

I can provide more details if needed.

**Comment by ryanbuening** [comment]
[Here is a blog post](https://blogs.iis.net/deanc/when-an-unhealthy-isapi-isn-t-really-unhealthy) from a Microsoft employee that seems somewhat related. They mention in the post that:

 _`...Even setting disallowRotationOnConfigChange to ‘true’ had no impact on the behavior.`_

**Comment by mus65** [comment]
Further things worth mentioning:

- Note that there two separate `disallowRotationOnConfigChange` settings, one for the ASP .NET Core Module as described [here](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/advanced?view=aspnetcore-8.0#disallow-rotation-on-config) and a separate config on the application pool itself. Since we were not sure which one is relevant in which case, we set both of these to true to be on the safe side.
- The `disallowRotationOnConfigChange` setting on the ASP .NET Core Module was only implemented with .NET 7, so the setting doesn't work if the installed module is too old.
- we have only seen this with .NET 8. Doesn't necessarily mean that this is a regression from .NET 6, we may just not have noticed it before.

**Comment by ryanbuening** [comment]
I have added a `web.config` file in the root of my Blazor Server application that gets deployed to IIS with the following config that sets `disallowRotationOnConfigChange` to `true`:

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <location path="." inheritInChildApplications="false">
    <system.webServer>
      <handlers>
        <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" />
      </handlers>
      <aspNetCore processPath="dotnet" arguments=".\AppName.dll" stdoutLogEnabled="false" stdoutLogFile=".\logs\stdout" hostingModel="inprocess">
        <handlerSettings>
          <handlerSetting name="disallowRotationOnConfigChange" value="true" />
        </handlerSettings>
      </aspNetCore>
    </system.webServer>
  </location>
</configuration>
```

I will report back what happens. We are/were seeing these random recycles at least once per day in our dev and test environment, so I should be able to tell if this change fixes things.

**Comment by ryanbuening** [comment]
Just wanted to follow up. I haven't experienced random app pool recycles since implementing the `web.config` change to `disallowRotationOnConfigChange` above. I will report back if I do.

**Comment by ryanbuening** [comment]
Wanted to follow up. Setting `disallowRotationOnConfigChange` definitely has fixed our issues with the app pool recycling randomly. My questions are:

1. Is it safe to keep this set to false? What are the negative repercussions?
2. Is this something that Microsoft is looking to address? In .NET 8? .NET 9? or is this working as designed?

@amcasey

**Comment by amcasey** [comment]
Thanks for the update!  Those would be good questions for @BrennanConroy.

**Comment by ryanbuening** [comment]
@BrennanConroy any thoughts, or are these questions I need to direct elsewhere? Thanks.

**Comment by ryanbuening** [comment]
@amcasey is there another person/resource I could direct my question to? I'm just trying to determine if this is something that is working as intended, in which case we will need to update all of our .NET 8+ apps on IIS with this `web.config` change. Thanks.

**Comment by amcasey** [comment]
Nope, Brennan's the right person - we've just got a lot going on right now.  I'll remind him, but no promises.

**Comment by amcasey** [comment]
In the meantime, I imagine you've seen this? https://learn.microsoft.com/en-us/previous-versions/iis/6.0-sdk/ms525764(v=vs.90)

**Comment by ryanbuening** [comment]
> In the meantime, I imagine you've seen this? https://learn.microsoft.com/en-us/previous-versions/iis/6.0-sdk/ms525764(v=vs.90)

So, according to the Configurable Locations in that link, it looks like the `DisallowRotationOnConfigChange` can be set globally for all application pools at `/LM/W3SVC/AppPools` which will be useful if the behavior described in this issue is the new expected behavior. Thanks.

**Comment by nshampur** [comment]
We are observing similar behavior

**Comment by BrennanConroy** [comment]
> Setting disallowRotationOnConfigChange definitely has fixed our issues with the app pool recycling randomly. My questions are:

> Is it safe to keep this set to false? What are the negative repercussions?

Yes it's safe. The repercussions are that your app won't automatically restart to get updates from your config. Some arbitrary examples would be modifying SSL settings, or adding IIS modules, or changing accepted mime types, etc.

> Is this something that Microsoft is looking to address? In .NET 8? .NET 9? or is this working as designed?

I don't know why you would get empty or null paths, but it seems like it would be a good idea to ignore empty paths. We already have a check for null so someone in the past must have realized that this sort of problem could occur.

**Comment by ryanbuening** [comment]
So is this something that can possibly be fixed by a change in the aspnetcore IIS module or will we need to continue to handle by configuring `disallowRotationOnConfigChange`?

**Comment by ryanbuening** [comment]
Coming back to this. I'm not sure what changed, but I would no longer recommend setting `disallowRotationOnConfigChange` to `true`. My Blazor Server application is still targeting .NET 8, but our admins installed .NET 9 on our Windows/IIS server. (I'm not sure if this has anything to do with it - or what else changed)

What I am seeing now with `disallowRotationOnConfigChange` set to `true` is when I see a log for OnGlobalConfigurationChange, the app will stop and get restarted, resulting in a 503 - Service Unavailable.

I can reproduce this on-demand by adding/removing a MIME Type in IIS Manager. Removing or setting `disallowRotationOnConfigChange` to `false` allows the application to at least restart - It does get recycled though.

```yml
[2024-12-11T20:40:36.805Z, PID: 9096] [aspnetcorev2.dll] ASPNET_CORE_GLOBAL_MODULE::OnGlobalConfigurationChange 'MACHINE/WEBROOT/APPHOST'
[2024-12-11T20:40:43.271Z, PID: 9096] [aspnetcorev2_inprocess.dll] Removing Request 0
[2024-12-11T20:40:43.272Z, PID: 9096] [aspnetcorev2.dll] ASPNET_CORE_GLOBAL_MODULE::OnGlobalApplicationStop
[2024-12-11T20:40:43.469Z, PID: 9096] [aspnetcorev2.dll] Shutdown starting in 1000 ms.
[2024-12-11T20:40:44.472Z, PID: 9096] [aspnetcorev2.dll] Shutdown starting.
[2024-12-11T20:40:44.473Z, PID: 9096] [aspnetcorev2.dll] Stopping application '/LM/W3SVC/7/ROOT'
[2024-12-11T20:40:44.475Z, PID: 9096] [aspnetcorev2_inprocess.dll] Stopping file watching.
[2024-12-11T20:40:44.486Z, PID: 9096] [aspnetcorev2_inprocess.dll] Waiting for file watcher thread to exit.
[2024-12-11T20:40:44.486Z, PID: 9096] [aspnetcorev2_inprocess.dll] Stopping file watcher thread
[2024-12-11T20:40:44.488Z, PID: 9096] [aspnetcorev2_inprocess.dll] Waiting for file watcher thread to exit.
[2024-12-11T20:40:44.489Z, PID: 9096] [aspnetcorev2_inprocess.dll] Stopping CLR
[2024-12-11T20:40:44.498Z, PID: 9096] [aspnetcorev2_inprocess.dll] Starting shutdown sequence 0
[2024-12-11T20:40:44.617Z, PID: 9096] [aspnetcorev2.dll] Received a request during shutdown. Will return a 503 response.
[2024-12-11T20:40:44.618Z, PID: 9096] [aspnetcorev2.dll] Failed HRESULT returned: 0x800704e7 at D:\a\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\AspNetCore\proxymodule.cpp:97 
[2024-12-11T20:40:44.619Z, PID: 9096] [aspnetcorev2.dll] Failed HRESULT returned: 0x800704e7 at D:\a\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\AspNetCore\proxymodule.cpp:121 
[2024-12-11T20:40:46.385Z, PID: 9096] [aspnetcorev2.dll] Received a request during shutdown. Will return a 503 response.
[2024-12-11T20:40:46.398Z, PID: 9096] [aspnetcorev2.dll] Failed HRESULT returned: 0x800704e7 at D:\a\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\AspNetCore\proxymodule.cpp:97 
[2024-12-11T20:40:46.401Z, PID: 9096] [aspnetcorev2.dll] Failed HRESULT returned: 0x800704e7 at D:\a\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\AspNetCore\proxymodule.cpp:121 
[2024-12-11T20:40:47.121Z, PID: 9096] [aspnetcorev2_inprocess.dll] Waiting for 0 requests to drain
[2024-12-11T20:40:47.122Z, PID: 9096] [aspnetcorev2_inprocess.dll] Drained all requests, notifying managed.
[2024-12-11T20:40:47.253Z, PID: 9096] [aspnetcorev2.dll] Received a request during shutdown. Will return a 503 response.
[2024-12-11T20:40:47.254Z, PID: 9096] [aspnetcorev2.dll] Failed HRESULT returned: 0x800704e7 at D:\a\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\AspNetCore\proxymodule.cpp:97 
[2024-12-11T20:40:47.258Z, PID: 9096] [aspnetcorev2.dll] Failed HRESULT returned: 0x800704e7 at D:\a\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\AspNetCore\proxymodule.cpp:121 
[2024-12-11T20:40:47.391Z, PID: 9096] [aspnetcorev2_inprocess.dll] Managed application exited with code 0
[2024-12-11T20:40:47.393Z, PID: 9096] [aspnetcorev2_inprocess.dll] Clr thread wait ended: clrThreadExited: 1
[2024-12-11T20:40:47.395Z, PID: 9096] [aspnetcorev2_inprocess.dll] Event Log: 'Application 'MACHINE/WEBROOT/APPHOST/DEVMYPOLICY' has shutdown.' 
End Event Log Message.
[2024-12-11T20:40:47.395Z, PID: 9096] [aspnetcorev2_inprocess.dll] Canceling standard stream pipe reader
[2024-12-11T20:40:47.397Z, PID: 9096] [aspnetcorev2_inprocess.dll] Stopping in-process worker thread
[2024-12-11T20:40:47.667Z, PID: 9096] [aspnetcorev2.dll] Received a request during shutdown. Will return a 503 response.
[2024-12-11T20:40:47.668Z, PID: 9096] [aspnetcorev2.dll] Failed HRESULT returned: 0x800704e7 at D:\a\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\AspNetCore\proxymodule.cpp:97 
[2024-12-11T20:40:47.669Z, PID: 9096] [aspnetcorev2.dll] Failed HRESULT returned: 0x800704e7 at D:\a\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\AspNetCore\proxymodule.cpp:121 
```

![Image](https://github.com/user-attachments/assets/ca2b21d5-e53d-464d-b485-daf8d7bbebe2)

**Comment by BrennanConroy** [comment]
That's a bug in the global module we need to fix. To workaround it and get back to 8.0 behavior, you can set `shutdownDelay` to `0`, either through the web.config setting shown below or the environment variable `ANCM_shutdownDelay`
```
<aspNetCore processPath="dotnet"
    arguments=".\MyApp.dll"
    stdoutLogEnabled="false"
    stdoutLogFile="\\?\%home%\LogFiles\stdout"
    hostingModel="inprocess">
  <handlerSettings>
    <handlerSetting name="shutdownDelay" value="0" />
  </handlerSettings>
</aspNetCore>
```

**Comment by BrennanConroy** [comment]
Null check fixed in https://github.com/dotnet/aspnetcore/pull/59998

**Comment by NGemballa** [comment]
Hi!

Will this be fixed for .NET 8 and 10, too? The workaround does not apply for us.

**Comment by BrennanConroy** [comment]
Can you try out the hosting bundle from 11.0-preview2 to see if the issues are resolved and there aren't any other problems? https://dotnet.microsoft.com/en-us/download/dotnet/11.0 using the following command line options to only install ANCM:
`OPT_NO_RUNTIME=1 OPT_NO_SHAREDFX=1 OPT_NO_X86=1` see https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/hosting-bundle?view=aspnetcore-10.0#options

**Comment by Youssef1313** [comment]
Moved to p4 as p2/p3 shipped already.

**Comment by ryanbuening** [comment]
@BrennanConroy, are these fixes only for .NET 11 apps?

**Comment by BrennanConroy** [comment]
> are these fixes only for .NET 11 apps?

Technically no. You need to install the 11.0 preview hosting bundle to get the global module installed, the file in `C:\Program Files\IIS\Asp.Net Core Module\V2`, but your actual app can be any version.

I am interested in backporting, but would love more people to try out the fixes, especially checking that the fixes don't introduce any new regressions (since we already introduced a regression which required these fixes). See https://github.com/dotnet/aspnetcore/issues/54531#issuecomment-4049258111 for installation details.

**Comment by pledwon2** [comment]
As requested, I want to leave a comment here regarding this issue: [AspnetcoreModuleV2 shutdowns worker threads but not worker process, which results in 503
](https://github.com/dotnet/aspnetcore/issues/66958)

The fix helped us with a problem which started occurring on net9+, where shutdownDelay was not 0 by default. 

I unfortunately can't provide any other info regarding how the fix affects other scenarios, as we use the workaround of setting shutdownDelay to 0, so we will not be using anything other that version 10 until LTS period ends.

**Comment by rojisamuel** [comment]
> Hi!
> 
> Will this be fixed for .NET 8 and 10, too? The workaround does not apply for us.

Please could we have a fix for this in .NET 10 too, please ?

**Comment by danroth27** [comment]
@rojisamuel @frulfump @JackCoplandGranta  To help us validate that this fix is appropriate to backport to .NET 10, can you please test if the fix in the latest .NET 11 preview resolves the issue for you? Your validation will help us unblock the backporting and servicing process.

**Comment by JackCoplandGranta** [comment]
@danroth27 Yes, Upgrading to the .Net 11 preview resolved the issue for us. Our App Pools are no longer left in a bad state.

**Comment by Frulfump** [comment]
I no longer work on a project with this issue so can't confirm.

**Comment by sghatbale** [comment]
@danroth27 
can we have the fix in .net10 too?

**Comment by danroth27** [comment]
@sghatbale It's under consideration. Have you already tried the fix in the latest .NET 11 preview? Can you confirm that the fix works for you?
